src/app/components/common/treenode.ts

```diff
@@ -1,14 +1,14 @@
-export interface TreeNode {
+export interface TreeNode<T = any> {
     label?: string;
-    data?: any;
+    data?: T;
     icon?: any;
     expandedIcon?: any;
     collapsedIcon?: any;
-    children?: TreeNode[];
+    children?: TreeNode<T>[];
     leaf?: boolean;
     expanded?: boolean;
     type?: string;
-    parent?: TreeNode;
+    parent?: TreeNode<T>;
```

Comment:
Im concerned with this one and the children one. Why would the parent node be of the same type?
Can't we have a `TreeNode<MyFile>` as a child of a `TreeNode<MyFolder>`?
Having the parent/child node types in the generic signature is probably cumbersome, so maybe it should be left as `any`

Reply:
Or maybe add the parent/child type in the generic signature, defaulting to the node item:
`TreeNode<T = any, P = T, C = T>`.

Reply:
With this change, your use case is left undisturbed (just make it `TreeNode<any>` -- or just `TreeNode`). It will just allow for stricter typing control for those who would like it/need it. 

Your other multi-typed generic proposal could be interesting! I'd like to see it. Make a PR! :)

Reply:
Ok. I was afraid that implementing this interface with different types would not compile, but it seems to compile just fine. Something like
`class MyNodeA implements TreeNode { value: string; children: MyNodeB[]; /**...**/}`
`class MyNodeB implements TreeNode { value: number; children: MyNodeB[]; /**...**/}`

Reply:
To the original example, you could also:

```ts
abstract class MyFileSystemNode { ... }
class MyFolder extends MyFileSystemNode { ... }
class MyFile extends MyFileSystemNode { ... }
```

Then a `TreeNode<MyFileSystemNode>` could have value of `MyFolder` and have child with value of `MyFile`.